File: roots/backend/_dict_harmonize.js
export const meta = {
  name: 'dict-harmonize',
  description: 'Harmonize scraped classical-dictionary root entries into a faithful translation_en + a readable harmonized_en (Qur’an-only voice). Draft-only + paced for usage economy; admin review is the faithfulness gate.',
  phases: [{ title: 'Draft' }],
}

// args = array of dictionary_entries.id to harmonize. The big raw Arabic stays in
// the dump file; each subagent reads ONLY its own entry by id (keeps prompts small).
const IDS = Array.isArray(args) ? args : JSON.parse(args)
const DUMP = '/Users/jawadshuaib/Desktop/projects/quran-related/roots/backend/data/dict_harmonize_dump.json'

const READ = (id) =>
  `python3 -c "import json; d=[r for r in json.load(open('${DUMP}',encoding='utf-8')) if r['id']==${id}][0]; ` +
  `print('ROOT', d['root_buckwalter'], d['root_arabic']); ` +
  `print('DICTIONARY', d['name_en'], '—', d['author'], '(d.'+str(d['author_death_year'])+' CE)'); ` +
  `print('LANGUAGE', d['language'], '| QURAN_SPECIFIC', d['is_quran_specific']); ` +
  `print('=== ORIGINAL ==='); print(d['original_text_ar'])"`

const DRAFT_SCHEMA = {
  type: 'object', additionalProperties: false,
  properties: {
    id: { type: 'integer' },
    translation_en: { type: 'string' },
    harmonized_en: { type: 'string' },
    confidence: { type: 'number' },
    issues: { type: 'array', items: { type: 'string' } },
  },
  required: ['id', 'translation_en', 'harmonized_en', 'confidence', 'issues'],
}

const VERIFY_SCHEMA = {
  type: 'object', additionalProperties: false,
  properties: {
    id: { type: 'integer' },
    ok: { type: 'boolean' },
    faithful: { type: 'boolean' },
    no_codified: { type: 'boolean' },
    complete: { type: 'boolean' },
    attribution_ok: { type: 'boolean' },
    severity: { type: 'string', enum: ['none', 'minor', 'major'] },
    reason: { type: 'string' },
  },
  required: ['id', 'ok', 'faithful', 'no_codified', 'complete', 'attribution_ok', 'severity', 'reason'],
}

function draftPrompt(id) {
  return `You are harmonizing ONE classical Arabic dictionary's entry for a single Qur'anic ROOT, for an internal scholarly reference on al-nuqta.com. The end goal is understanding the Qur'an's own language, so meaning is derived from attested usage — never from later codified doctrine.

STEP 1 — read the exact entry (run this):
\`\`\`
${READ(id)}
\`\`\`

STEP 2 — from the ORIGINAL ARABIC ABOVE ONLY (do not add senses from memory or from other dictionaries), produce TWO English renderings of THIS one entry:

(A) translation_en — a FAITHFUL, close translation. Keep it complete for the lexical substance: every distinct sense, the grammatical/morphological notes (wazn, iʿrāb, plural patterns), the derivation/etymology, and the poetic or Qur'anic shawāhid (citations) with their attributions. You MAY drop pure transmission chains ("A told us from B…") and collapse verbatim repetition, but never summarise a sense away. Give Arabic head-words/examples in Arabic followed by a parenthesised English gloss.

(B) harmonized_en — the SAME content made READABLE for a serious non-Arabist. Clear modern English, as short sense-numbered points or brief paragraphs. KEEP every sense-distinction, the grammatical method, the etymology, and the key shawāhid. REMOVE: isnād chains, piles of near-synonyms, tangents, and edition/scribal apparatus. Trim length, but do NOT flatten distinct senses into one or lose nuance. For a very long entry (e.g. Lisān), this is a faithful digest, not a paraphrase of the first line.

VOICE + INTEGRITY — hard rules:
- Present this as THIS lexicographer's account, described neutrally as evidence. No advocacy, no first person ("we"), no warning the reader toward or away from a reading.
- Meaning comes from the attested usage this entry records. Do NOT import later codified juristic/theological senses the Arabic does not state (do not gloss ṣalāt as "the five daily ritual prayers", zakāt as "the alms-tax", ḥajj as "the pilgrimage rite", īmān as a creed) UNLESS the entry itself says so — and then attribute it to this author, not as fact.
- Faithfulness over completeness: every clause in BOTH renderings must trace to the Arabic above. If the entry is terse, your renderings are terse. Never pad, never invent a citation.
- If the source is already English (Lane, Salmoné): translation_en = the original lightly cleaned; harmonized_en = a readable condensation.
- BOTH fields must be REAL, complete outputs — NEVER write "placeholder", "unused", "see above", or stub a field. For a very long entry, condense faithfully (drop repeated shawāhid + isnād) so both stay a reasonable length, but each must carry genuine content.

Return id=${id}, translation_en, harmonized_en, confidence (0–1: how faithful+complete you judge your renderings), and issues (array; note any truncation/garbling/uncertainty, else []).`
}

function verifyPrompt(id, draft) {
  return `You are the FAITHFULNESS CHECKER for a harmonized classical-dictionary entry. Be strict and adversarial: your job is to catch invented meaning, imported doctrine, flattened senses, and editorializing.

STEP 1 — read the ORIGINAL (run this):
\`\`\`
${READ(id)}
\`\`\`

STEP 2 — judge this DRAFT against the original:
--- translation_en ---
${draft.translation_en}
--- harmonized_en ---
${draft.harmonized_en}

Checks (each true = passes):
- faithful: every sense/claim/citation in BOTH renderings traces to the original Arabic; nothing invented, no citation fabricated.
- no_codified: the draft did NOT import post-Qur'anic codified juristic/theological definitions that the original does not itself state.
- complete: the original's main sense-distinctions are all present (not collapsed into one); grammatical/etymological substance retained where the original has it.
- attribution_ok: neutral evidence voice — no advocacy, no first person, no reader-warnings; reads as this author's account.

ok = (faithful AND no_codified AND complete AND attribution_ok). severity = 'major' if faithful or no_codified failed, 'minor' if only complete/attribution slipped, else 'none'. reason = one line citing the specific problem (or "clean"). Return id=${id}.`
}

log(`Harmonizing ${IDS.length} dictionary entries (draft-only, paced)`)
const results = await parallel(
  IDS.map((id) => () => agent(draftPrompt(id), { schema: DRAFT_SCHEMA, label: `draft:${id}`, phase: 'Draft' }))
)
const out = results.filter(Boolean)
log(`Drafted ${out.length}/${IDS.length}`)
return out

File: roots/backend/dict_harmonize.py
#!/usr/bin/env python3
"""Harmonization prep for the Lexicon Library.

`prep` dumps the PENDING `dictionary_entries` that still need harmonization
(no `harmonized_en` yet) to a JSON worklist that the `_dict_harmonize.js`
workflow reads per-entry (by id, via the dump path) — keeping the big raw Arabic
out of prompts/args. The workflow drafts translation_en + harmonized_en, a verify
stage checks faithfulness, and `_apply_dict_drafts.py` writes the result back
(still `pending`) for admin review.

Usage:
    python3 dict_harmonize.py prep Amn qwl        # dump pending entries for these roots
    python3 dict_harmonize.py prep --top 300      # dump pending for top-300 frequency roots
    python3 dict_harmonize.py prep --limit 40     # dump first 40 pending entries
    python3 dict_harmonize.py prep --redo Amn     # re-dump even already-harmonized entries
    python3 dict_harmonize.py status              # harmonization progress
"""
import json
import os
import sqlite3
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
DB = os.path.join(HERE, "data", "quran.db")
DUMP = os.path.join(HERE, "data", "dict_harmonize_dump.json")   # the workflow reads THIS


def _conn():
    c = sqlite3.connect(DB, timeout=60)
    c.row_factory = sqlite3.Row
    return c


def _top_roots(conn, n):
    return [r[0] for r in conn.execute(
        "SELECT root_buckwalter FROM morphology WHERE root_buckwalter IS NOT NULL "
        "AND root_buckwalter!='' GROUP BY root_buckwalter ORDER BY COUNT(*) DESC LIMIT ?",
        (n,)).fetchall()]


def prep(roots=None, limit=None, redo=False):
    conn = _conn()
    q = ("SELECT e.id, e.root_buckwalter, e.root_arabic, e.dictionary_slug, "
         "e.original_text_ar, d.name_en, d.author, d.author_death_year, "
         "d.is_quran_specific, d.language "
         "FROM dictionary_entries e JOIN dictionaries d ON d.slug = e.dictionary_slug "
         "WHERE e.original_text_ar IS NOT NULL AND LENGTH(e.original_text_ar) >= 3 ")
    args = []
    if not redo:
        q += "AND (e.harmonized_en IS NULL OR e.harmonized_en = '') "
    if roots:
        q += "AND e.root_buckwalter IN (%s) " % ",".join("?" * len(roots))
        args += roots
    q += "ORDER BY e.root_buckwalter, d.author_death_year"
    if limit:
        q += " LIMIT ?"
        args.append(limit)
    rows = [dict(r) for r in conn.execute(q, args).fetchall()]
    conn.close()
    with open(DUMP, "w", encoding="utf-8") as f:
        json.dump(rows, f, ensure_ascii=False)
    ids = [r["id"] for r in rows]
    print(json.dumps({"dump": DUMP, "count": len(rows), "roots": sorted({r["root_buckwalter"] for r in rows}),
                      "ids": ids}, ensure_ascii=False))
    return ids


def next_chunk(n):
    """Dump ALL still-pending entries to the worklist and print the next N ids
    (frequency-ordered) as a JSON array — one paced /loop tick's batch."""
    conn = _conn()
    freq = {r[0]: r[1] for r in conn.execute(
        "SELECT root_buckwalter, COUNT(*) FROM morphology WHERE root_buckwalter!='' "
        "GROUP BY root_buckwalter").fetchall()}
    rows = [dict(r) for r in conn.execute(
        "SELECT e.id, e.root_buckwalter, e.root_arabic, e.dictionary_slug, e.original_text_ar, "
        "d.name_en, d.author, d.author_death_year, d.is_quran_specific, d.language "
        "FROM dictionary_entries e JOIN dictionaries d ON d.slug = e.dictionary_slug "
        "WHERE e.original_text_ar IS NOT NULL AND LENGTH(e.original_text_ar) >= 3 "
        "AND (e.harmonized_en IS NULL OR e.harmonized_en = '')").fetchall()]
    conn.close()
    with open(DUMP, "w", encoding="utf-8") as f:
        json.dump(rows, f, ensure_ascii=False)
    # frequency-ordered so the most load-bearing roots harmonize first; whole
    # roots kept together within the chunk for a coherent panel.
    rows.sort(key=lambda r: (-freq.get(r["root_buckwalter"], 0), r["root_buckwalter"], r["author_death_year"] or 0))
    ids = [r["id"] for r in rows[:n]]
    print(json.dumps({"remaining": len(rows), "chunk": ids}, ensure_ascii=False))


def status():
    conn = _conn()
    tot = conn.execute("SELECT COUNT(*) FROM dictionary_entries").fetchone()[0]
    done = conn.execute("SELECT COUNT(*) FROM dictionary_entries WHERE harmonized_en IS NOT NULL AND harmonized_en!=''").fetchone()[0]
    appr = conn.execute("SELECT COUNT(*) FROM dictionary_entries WHERE review_status='approved'").fetchone()[0]
    roots = conn.execute("SELECT COUNT(DISTINCT root_buckwalter) FROM dictionary_entries").fetchone()[0]
    print(f"entries: {tot}  harmonized: {done}  approved: {appr}  distinct roots: {roots}")
    conn.close()


def main():
    cmd = sys.argv[1] if len(sys.argv) > 1 else "help"
    if cmd == "prep":
        rest = sys.argv[2:]
        redo = "--redo" in rest
        rest = [a for a in rest if a != "--redo"]
        limit = None
        roots = None
        if "--top" in rest:
            i = rest.index("--top")
            conn = _conn(); roots = _top_roots(conn, int(rest[i + 1])); conn.close()
        elif "--limit" in rest:
            i = rest.index("--limit"); limit = int(rest[i + 1])
        else:
            roots = rest or None
        prep(roots=roots, limit=limit, redo=redo)
    elif cmd == "next":
        next_chunk(int(sys.argv[2]) if len(sys.argv) > 2 else 30)
    elif cmd == "status":
        status()
    else:
        print(__doc__)


if __name__ == "__main__":
    main()
